xdsl/dialects/onnx.py

```diff
@@ -0,0 +1,105 @@
+from __future__ import annotations
+
+from typing import Annotated, TypeVar
+
+from xdsl.dialects.builtin import (
+    SSAValue,
+    TensorType,
+)
+from xdsl.ir import (
+    Attribute,
+    Dialect,
+)
+from xdsl.irdl import (
+    ConstraintVar,
+    IRDLOperation,
+    irdl_op_definition,
+    operand_def,
+    result_def,
+)
+from xdsl.utils.exceptions import VerifyException
+
+_T = TypeVar("_T", bound=Attribute)
+
+
+class ElementwiseBinOpBase(IRDLOperation):
+    """Base class for element-wise binary operations on tensors with Numpy-style broadcasting."""
+
+    T = Annotated[Attribute, ConstraintVar("T"), _T]
+    lhs = operand_def(TensorType[T])
+    rhs = operand_def(TensorType[T])
+    res = result_def(TensorType[T])
+    assembly_format = "`(` $lhs `,` $rhs `)` attr-dict `:` `(` type($lhs) `,` type($rhs) `)` `->` type($res)"
+
+    def __init__(
+        self,
+        lhs: SSAValue,
+        rhs: SSAValue,
+    ):
+        super().__init__(
+            operands=[lhs, rhs],
+            result_types=[[]],
+        )
+
+    def verify_(self) -> None:
+        # Check that the arguments are broadcastable (using Numpy semantics) and that the result type is correct.
+        res_shape: list[int] = []
+
+        # Iterate over the shapes in reverse order and compute the result shape.
+        if (
+            not isinstance(lhs_type := self.lhs.type, TensorType)
+            or not isinstance(rhs_type := self.rhs.type, TensorType)
+            or not isinstance(res_type := self.res.type, TensorType)
+        ):
+            raise ValueError(
+                "onnx elementwise binary operation operands and result must be of type TensorType"
+            )
```

Comment:
I would make an `assert` for this here, as the `verify_` function is always called after the IRDL verifier, which will always ensure that these are `TensorType`.
The `assert` makes it clearer that this is for pyright only.